counsel.el

```diff
@@ -839,8 +857,10 @@ Optional INITIAL-INPUT is the initial input in the minibuffer."
               :history 'counsel-M-x-history
               :action
               (lambda (cmd)
-                (when (featurep 'smex)
-                  (smex-rank (intern cmd)))
+                (cond (smex-initialized-p
+                       (smex-rank (intern cmd)))
+                      (amx-initialized
+                       (amx-rank (intern cmd))))
```

Comment:
Looks like `(intern cmd)` is being used a lot in this function; might be worth `(setq cmd (intern cmd))` once and for all.

Reply:
Instead of relying on `smex-initialized-p` and `amx-initialized` to be defined, better check `(featurep 'amx)`/`(featurep 'smex)` or, for extra safety, `(fboundp 'amx-rank)`/`(fboundp 'smex-rank)`.

Reply:
About the second comment, the variables are not guaranteed by the code a couple of lines before? I just wanted to avoid extra 'relative heavy' functions calls when we could use a simple logical symbol.

Reply:
> the variables are not guaranteed by the code a couple of lines before?

The variables are only bound if you call `(defvar VAR VAL)`, but IMO it is bad to define variables from external packages. If the variables are declared via `(defvar VAR)`, as I am suggesting, then they are not guaranteed to be bound at this point if neither `amx` nor `smex` is installed. An alternative to the `featurep` and `fboundp` checks is `(bound-and-truep smex-initialized-p)`/`(bound-and-true-p amx-initialized)`. I have no preference between the three guards in this case; it's up to you and @abo-abo.

> I just wanted to avoid extra 'relative heavy' functions calls when we could use a simple logical symbol.

`featurep` and `fboundp` are far from being heavy, *especially* compared to the rest of the action function. Besides, this function only gets called once as part of a much larger interactive operation. There really is no argument to be made for performance here, especially if the argument is speculative/uninformed and not backed by profiling/benchmarking results.